### PR TITLE
Workspace nested dependency builds

### LINKED ISSUE #1800: --cache-builds always see an invalid cache under certain circumstances
* `carthage version`: 0.20.0
* `xcodebuild -version`: 8.2.1
* Are you using `--no-build`? no
* Are you using `--no-use-binaries`? no
* Are you using `--use-submodules`? no

**Cartfile**
```
github "Moya/Moya" "7.0.4"
```

**Carthage Output**
```
*** Fetching RxSwift
*** Fetching Alamofire
*** Fetching Result
*** Checking out Alamofire at "3.5.1"
*** Checking out Result at "2.1.3"
*** Checking out ReactiveCocoa at "v4.2.2"
*** Checking out RxSwift at "2.6.1"
*** Checking out Moya at "7.0.4"
*** xcodebuild output can be found in /var/folders/kk/vhmt6xwd5f17hfq_zz_lbnth0000gn/T/carthage-xcodebuild.3Q6zoJ.log
*** Valid cache found for Alamofire, skipping build
*** Invalid cache found for Result, rebuilding with all downstream dependencies
*** Valid cache found for RxSwift, skipping build
*** Building scheme "Result-iOS" in Result.xcodeproj
*** Building scheme "ReactiveCocoa-iOS" in ReactiveCocoa.xcworkspace
*** Building scheme "Moya iOS" in Moya.xcodeproj
*** Building scheme "ReactiveMoya iOS" in Moya.xcodeproj
*** Building scheme "RxMoya iOS" in Moya.xcodeproj
```

With this Cartfile, repeated runs of `carthage update --platform ios --cache-builds` detect an invalid cache for `Result` and rebuild it and it's dependents every time.

### PR DESCRIPTION
**Pull request by BobElDevil**
Fixes #1800 

If the source of the target is a project file within the given dependencies own carthage folder, do not copy it into the final build results folder

### CODE REVIEW

File: Source/CarthageKit/Xcode.swift [hunk 1]
```diff
@@ -449,11 +449,14 @@ public func buildScheme(_ scheme: String, withOptions options: BuildOptions, inP
...
-						if let frameworkType = settings.frameworkType.value {
-							return shouldBuildFrameworkType(frameworkType)
-						} else {
-							return false
+						if let frameworkType = settings.frameworkType.value, shouldBuildFrameworkType(frameworkType) {
+							if let projectPath = settings.projectPath.value {
+								let projectURL = URL(fileURLWithPath: projectPath)
+								let dependencyCheckoutDir = workingDirectoryURL.appendingPathComponent(CarthageProjectCheckoutsPath)
+								return !dependencyCheckoutDir.hasSubdirectory(projectURL)
+							}
```
Comment (jdhealy) [inline Source/CarthageKit/Xcode.swift]: Might be better to invert this, returning false with a `guard`…`else`:

```swift
guard
	settings.frameworkType.value.map(shouldBuildFrameworkType) == true,
	let projectURL = settings.projectPath.value.map(URL.init(fileURLWithPath:))
else {
	return false
}

let dependencyCheckoutsURL = workingDirectoryURL.appendingPathComponent(CarthageProjectCheckoutsPath, isDirectory: true)
return !dependencyCheckoutsURL.hasSubdirectory(projectURL)
```

File: Source/CarthageKit/Xcode.swift [hunk 2]
```diff
@@ -449,11 +449,14 @@ public func buildScheme(_ scheme: String, withOptions options: BuildOptions, inP
 				return BuildSettings.loadWithArguments(argsForLoading)
 					.filter { settings in
 						// Only copy build products for the framework type we care about.
```
Comment (jdhealy) [inline Source/CarthageKit/Xcode.swift]: Would you mind revising (along the lines of the commit message) this comment?

**Comment by jdhealy** [comment]
Like @miep, this seems very reasonable to me on a conceptual level — great job doing the investigatory and this PR!

The fixture test is great, but (and I think this shouldn’t be difficult) I'd rather have a smaller zip file than the 683 KB `ReactiveSwift-1.1.zip`, just adds too much to every clone.

Attached is a 71 KB `WorkspaceWithNestedDependencies.zip`, based off `DependencyTest.zip` with `TestFramework1` with a `Carthage/Checkouts`-housed `TestFramework2` where `…1`‘s scheme builds both.

A few other notes and then a strategy: 

- I'd prefer the commit message to be more precise, mentioning workspaces, schemes, and the `Carthage/Checkouts` directory. 

- See [this note about considering `guard`…`else`](https://github.com/Carthage/Carthage/pull/1831#discussion_r106071618).

*Strategy*: just to be extra safe in not merging in `ReactiveSwift-1.1.zip`, would you mind starting a new Pull Request, with “Allow edits from maintainers” checkbox checked. The first commit being the CarthageKit changes with the notes above addressed. Then, your test block-commented out (see below), but without the zip file.

Afterwards, I could commit myself — uncommenting the test, editing it, and adding the `WorkspaceWithNestedDependencies.zip` fixture.

I know it's an ask, but probably work it to save a half-MB on the git repo… 😄 

<details>
<summary>Test — Block Comment style</summary>

```swift
/*
it("should not copy build products from nested dependencies produced by workspace") {
	let _directoryURL = Bundle(for: type(of: self)).url(forResource: "ReactiveSwift-1.1", withExtension: nil)!
	let _buildFolderURL = _directoryURL.appendingPathComponent(CarthageBinariesFolderPath)
	
	_ = try? FileManager.default.removeItem(at: _buildFolderURL)
	
	let result = buildInDirectory(_directoryURL, withOptions: BuildOptions(configuration: "Debug", platforms: [.macOS]))
		.ignoreTaskData()
		.on(value: { (project, scheme) in
			NSLog("Building scheme \"\(scheme)\" in \(project)")
		})
		.wait()
	expect(result.error).to(beNil())
	
	let reactiveSwiftPath = _buildFolderURL.appendingPathComponent("Mac/ReactiveSwift.framework").path
	let resultPath = _buildFolderURL.appendingPathComponent("Mac/Result.framework").path
	
	expect(reactiveSwiftPath).to(beExistingDirectory())
	expect(resultPath).notTo(beExistingDirectory())
}
*/
```

</details>

<br/>

[WorkspaceWithNestedDependencies.zip](https://github.com/Carthage/Carthage/files/843217/WorkspaceWithNestedDependencies.zip)

**Comment by BobElDevil** [comment]
@jdhealy Made the requested changes. I actually had those test frameworks sitting around still from when I created DependencyTest, so I was able to create the workspace set up, but cleaned it up a bit more by removing redundant targets and unnecessary .git folders. the zip is now 25k.

I just rebased and did a force push so the larger zip isn't in the history at all, seemed easier than playing hot potato with the PR =)

File: Source/CarthageKit/Xcode.swift [hunk 3]
```diff
@@ -448,12 +448,15 @@ public func buildScheme(_ scheme: String, withOptions options: BuildOptions, inP
...
-							return shouldBuildFrameworkType(frameworkType)
-						} else {
+						// Only copy build products that are dynamic frameworks
+						guard let frameworkType = settings.frameworkType.value, shouldBuildFrameworkType(frameworkType), let projectPath = settings.projectPath.value else {
 							return false
 						}
+
+						// Do not copy build products that originate from the current project's own carthage dependencies
+						let projectURL = URL(fileURLWithPath: projectPath)
+						let dependencyCheckoutDir = workingDirectoryURL.appendingPathComponent(CarthageProjectCheckoutsPath)
```
Comment (jdhealy) [inline Source/CarthageKit/Xcode.swift]: :+1: This all looks good, do you just mind adding in `isDirectory: true`?

File: Source/CarthageKitTests/XcodeSpec.swift [hunk 4]
```diff
@@ -256,6 +256,27 @@ class XcodeSpec: QuickSpec {
...
+			let result = buildInDirectory(_directoryURL, withOptions: BuildOptions(configuration: "Debug", platforms: [.macOS]))
+				.ignoreTaskData()
+				.on(value: { (project, scheme) in
+					NSLog("Building scheme \"\(scheme)\" in \(project)")
+				})
+				.wait()
+			expect(result.error).to(beNil())
+
+			let reactiveSwiftPath = _buildFolderURL.appendingPathComponent("Mac/TestFramework1.framework").path
+			let resultPath = _buildFolderURL.appendingPathComponent("Mac/TestFramework2.framework").path
```
Comment (jdhealy) [inline Source/CarthageKitTests/XcodeSpec.swift]: Gardening: rename these variables 🌱

**Comment by mdiep** [comment]
Great work! Thanks. ✨

**Pull request merged**